godot-macros/src/util/list_parser.rs

```diff
@@ -0,0 +1,206 @@
+        };
+
+        if tokens.len() > 1 {
+            return bail!(&tokens[1], "unexpected expression");
+        }
+
+        Ok(Some(Self::new_from_tree(tokens.remove(0), delimiter)?))
+    }
+
+    /// Create a new parser from the given tokentree.
+    ///
+    /// Ensures that the tree is a list of lists of tokentrees, delimited by the provided delimiter, where
+    /// no sublist is empty. Except for the last list, which is allowed to be empty to allow for trailing
+    /// commas.
+    pub fn new_from_tree(tree: TokenTree, delimiter: Delimiter) -> ParseResult<Self> {
+        let group = match tree {
+            TokenTree::Group(group) => group,
+            _ => return bail!(tree, "expected list of items"),
+        };
+
+        if group.delimiter() != delimiter {
+            let expected = delimiter_opening_char(delimiter);
+            let got = delimiter_opening_char(group.delimiter());
+
+            return bail!(group.span_open(), "expected `{expected}`, got `{got}`");
+        }
+
+        let trees: Vec<TokenTree> = group.stream().into_iter().collect();
+
+        let raw_lists = trees
+            .split_inclusive(|tree| is_punct(tree, ','))
+            .collect::<Vec<_>>();
+
+        let list_len = raw_lists.len();
+        let mut lists = Vec::new();
+
+        for (i, list) in raw_lists.into_iter().enumerate() {
+            let is_last = i == list_len - 1;
+
+            // every list except the last one must contain at least a comma so this means we have a trailing
+            // comma.
+            if list.is_empty() {
+                break;
+            }
+
+            // does list only contain `,`?
+            if !is_last && list.len() == 1 {
+                let list_stream = list.iter().cloned().collect::<TokenStream>();
+                return bail!(list_stream, "expected expression, found `,`");
+            }
+
+            let end = if is_last { list.len() } else { list.len() - 1 };
+
+            lists.push(KvValue::new((&list[..end]).into()));
+        }
+
+        Ok(Self {
+            lists: lists.into(),
+            span_close: group.span_close(),
+        })
+    }
+
+    /// Get the next element from the list, starting at the front.
+    fn pop_next(&mut self) -> Option<KvValue> {
+        self.lists.pop_front()
+    }
+
+    fn peek(&self) -> Option<&KvValue> {
+        self.lists.front()
+    }
+
+    /// Take the next element of the list, ensuring it is an expression.
+    pub fn next_expr(&mut self) -> ParseResult<TokenStream> {
+        let Some(kv) = self.pop_next() else {
+            return bail!(self.span_close, "expected expression");
+        };
+
+        kv.expr()
+    }
+
+    /// Take the next element of the list, if it is an identifier.
+    ///
+    /// Returns `Ok(None)` if there are no more elements left.
+    pub fn next_ident(&mut self) -> ParseResult<Option<Ident>> {
+        let Some(kv) = self.pop_next() else { return Ok(None) };
```

Comment:
`let/else` should have the `else` branch on its own line, so it's very visible that a return is happening. It's also consistent with `if` formatting.

(Yes, it's annoying that rustfmt makes no progress...)

Reply:
current rustfmt master formats this let-else on one line

Reply:
As agreed on Discord, looks like future rustfmt will address `let/else` and format it the same way as is here, so no need to change.